### ISSUE TITLE
"There was an error connecting to UserVoice" on UserVoice.launchUserVoice()

### ISSUE THREAD

**Issue by couchcrew-thomas** [uservoice/uservoice-android-sdk#170]
Whenever I'm tryning to launch the uservoice activity, it instantly shows an error with the message "There was an error connecting to UserVoice" (it actually shows the german version of this).

I am using the latest version of the SDK and itegrated it using those lines:

In Application.onCreate:

    Config config = new Config(getString(R.string.uservoice_site));
    config.setForumId(Integer.parseInt(getString(R.string.uservoice_forumid)));
    UserVoice.init(config, this);

When the activity should be launched:

    UserVoice.launchUserVoice(getActivity());

The log shows the following three messages:

    D/UV﹕ /api/v1/clients/default.json
    D/UV﹕ https://by.uservoice.com/t/k/redacted/d/k/track.js?_=1424078715580&c=_
    E/com.uservoice.uservoicesdk﹕ Host name may not be null -- null

According to my research this error occurs if HTTP requests with underscores inside them are send using the apache commons http lib.

Does anyone know a solution? We'll have to abandon UV if there is no fix during the next days.

Edit: We're currently falling back to opening the forum in the device's browser. This works for the moment but is no permanent solution.

**Comment by austintaylor** [comment]
As far as I can tell, people are getting that error when they have underscores in their *host name* which is invalid, and is not the case here. You can have underscores in params.

Are you using the latest version of the sdk? There was a bug in older versions where the Config itself would get lost if the Android process had been restarted, which could cause the host name to be lost. This was fixed in 1.2.1.

**Comment by couchcrew-thomas** [comment]
Im using the latest version thats available from the maven repository.

    compile 'com.uservoice:uservoice-android-sdk:+'

**Comment by austintaylor** [comment]
What is the actual string you are passing to the config? I cannot reproduce the error.

**Comment by couchcrew-thomas** [comment]
I'm passing the site name of our uservoice:

        <string name="uservoice_site">couchfunk.uservoice.com​</string>

**Comment by austintaylor** [comment]
So there's nothing wrong with the way you have it set up, and if I test it locally, it works perfectly. Are you able to step through it in a debugger to see what is going on?

All I can tell from what you've told me is that you are passing in the site URL, but by the time it goes to talk to the UserVoice API, that URL is null. In the past this would happen if the app process got pushed out of memory after the sdk was initialized, but before it was actually launched. However, that has been fixed in the latest version, which you are using, by saving the Config in local storage. (It was also always an intermittent bug, whereas it sounds like you are getting the same error message every time you try to launch the sdk.)

Is there any other information you can give me about what it is doing? There is not much I can do without being able to reproduce the issue.

**Comment by ChrisMCMine** [comment]
Same happens for me when a user tries to post a new idea.

A good workaround for this case, to still receive a new idea or a new ticket, would be to open the uservoice portal in the browser or in an in-app webview when this error occurs.

**Comment by ChrisMCMine** [comment]
Any update on this?

**Comment by couchcrew-thomas** [comment]
No. It's still not working.

I'm currently occupied with non Android tasks.
When I have some time, I will create a blank app with only uservoice and
try the integration.
If it works, I'll add our Apps' depencendy chain to this app. My guess is
still that it somehow clashes with another dependency.
That would explain why it occurs in all our apps and you cannot reproduce
it on your side.

2015-04-14 18:52 GMT+02:00 ChrisMCMine <notifications@github.com>:

> Any update on this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/uservoice/uservoice-android-sdk/issues/170#issuecomment-92976357>
> .
>



-- 
Couchfunk - Unterhaltung, die mitreißt!
Fernsehen, Sport-Entertainment, Webvideo

www.Couchfunk.de <http://www.couchfunk.de/>
www.TV.de <http://www.tv.de/>

Mobil: 0176-61798686
Tel: 0351-49764380
Fax: 0351-49764388

Anschrift:
Couchfunk GmbH
Wichernstr. 10, 01445 Radebeul
Handelsregister: Amtsgericht Dresden HRB 30422
Geschäftsführer: Uz Kretzschmar

**Comment by nJackx** [comment]
I'm having the same problem. 
I have a button in my app the user can use to open the uservoice panel, but 90% of the times the message "there was a problem connecting to uservoice" pops up.

My code to initialize uservoice is this:

    Config config = new Config("xyz.uservoice.com");
    config.identifyUser(userId, name, email);

    UserVoice.init(config, this);
    UserVoice.launchContactUs(this);

**Comment by austintaylor** [comment]
I found some more information about this. Unfortunately, I don't yet have a fix.

I was able to reproduce this by launching the sdk in a newly created virtual device. It still does not happen on every 1st launch, but it happens frequently enough that I could catch it in the debugger. The issue is that it is getting a 304 Not Modified response from the server when requesting /api/v1/clients/default.json. The null pointer exception is coming from the JSON library trying to parse a null response body. One thing we could do to improve the sdk would be to check for null in that code and log it explicitly so that this kind of bug is easier to diagnose in the future.

The 304 response does not make sense to me, and my current theory is that we have a bug in our server-side caching logic. (I am not sending If-None-Match or If-Modified-Since, unless something in Android is adding it for me. I also can't see any way that it could have a valid etag since the bug happens on first launch.)

My next step will be to capture the requests with Wireshark to see if that yields any more insight as to why the server is returning a 304. So far I have not been able to produce this response just by hitting the endpoint with curl.
